[10.0][ADD] report_paper_wkhtmltopdf_params

JordiBForgeFlow opened the pull request:
Report Paper Wkhtmltopdf Params
===========================

This module allows you to add new parameters for a paper format which are
then forwarded to wkhtmltopdf command as arguments.

Usage
=====

* Go to *Settings* and press 'Activate the developer mode (with assets)'

* Go to *Settings - Technical - Reports - Paper Format*

* Change the parameter 'Disable Smart Shrinking'.

* Add additional parameters indicating the command argument name (remember to
  add prefix -- or -) and value.

Commit:
[ADD] module report_paper_wkhtmltopdf_params

JordiBForgeFlow commented:
This is a migration of the module created by @mlaitinen, from https://github.com/avoinsystems/report_paper_params

pedrobaeza commented:
You should credit him in the commit author.

mlaitinen commented:
Thanks @jbeficent for creating the PR. I've been extremely busy lately and this has been on my todo list.

JordiBForgeFlow commented:
@pedrobaeza Done! @mlaitinen Thanks to you!! That's what we're here for, to help each other :)

elicoidal commented on report_paper_wkhtmltopdf_params/README.rst:
```diff
@@ -0,0 +1,65 @@
+===============================
+
+This module allows you to add new parameters for a paper format which are
+then forwarded to wkhtmltopdf command as arguments.
```
Can you explain the param you are tweaking and the advantage of modifying it?

elicoidal commented on report_paper_wkhtmltopdf_params/README.rst:
```diff
@@ -0,0 +1,65 @@
+Usage
+=====
+
+* Go to *Settings* and press 'Activate the developer mode (with assets)'
```
usually better enumerated list for procedures ("#. ")

elicoidal commented on report_paper_wkhtmltopdf_params/README.rst:
```diff
@@ -0,0 +1,65 @@
+* Change the parameter 'Disable Smart Shrinking'.
+
+* Add additional parameters indicating the command argument name (remember to
+  add prefix -- or -) and value.
```
Can you provide an example and/or a link to the param list?

elicoidal commented on report_paper_wkhtmltopdf_params/README.rst:
```diff
@@ -0,0 +1,65 @@
+
+* Go to *Settings - Technical - Reports - Paper Format*
+
+* Change the parameter 'Disable Smart Shrinking'.
```
Why smart shrinking is so special? (I know: it fitted your needs)
My point is why having a module for this special attribute and allow then to have another free text for whatever attributes you want to add.
I would rather have it without the checkbox for smart-shrinking (then you just have to set up the module for your own usage) or with all options available for any (or most of) the good options that wkhtml2pdf can offer.
Option 1 is more flexible and lighter and more prone to errors
Option 2 avoids Odoo crashing controlling the input (which somehow I kinda prefer)

What do you think?

JordiBForgeFlow replied:
Will properly update documentation. smart shrinking is important when you intend to have a layout that conforms to certain alignment. It is very common whenever you need to conform the PDF to a predefined layoyut (e.g. checks, official forms,...).

It is specifically proposed because it is the most common use case where you need to add a parameter to wkhtmltopdf.

As for the free parameters, they are only supposed to be used by admin users.

The options available are:
`Global Options:
      --collate                       Collate when printing multiple copies
                                      (default)
      --no-collate                    Do not collate when printing multiple
                                      copies
      --cookie-jar <path>             Read and write cookies from and to the
                                      supplied cookie jar file
      --copies <number>               Number of copies to print into the pdf
                                      file (default 1)
  -d, --dpi <dpi>                     Change the dpi explicitly (this has no
                                      effect on X11 based systems) (default 96)
  -H, --extended-help                 Display more extensive help, detailing
                                      less common command switches
  -g, --grayscale                     PDF will be generated in grayscale
  -h, --help                          Display help
      --htmldoc                       Output program html help
      --image-dpi <integer>           When embedding images scale them down to
                                      this dpi (default 600)
      --image-quality <integer>       When jpeg compressing images use this
                                      quality (default 94)
      --license                       Output license information and exit
  -l, --lowquality                    Generates lower quality pdf/ps. Useful to
                                      shrink the result document space
      --manpage                       Output program man page
  -B, --margin-bottom <unitreal>      Set the page bottom margin
  -L, --margin-left <unitreal>        Set the page left margin (default 10mm)
  -R, --margin-right <unitreal>       Set the page right margin (default 10mm)
  -T, --margin-top <unitreal>         Set the page top margin
  -O, --orientation <orientation>     Set orientation to Landscape or Portrait
                                      (default Portrait)
      --page-height <unitreal>        Page height
  -s, --page-size <Size>              Set paper size to: A4, Letter, etc.
                                      (default A4)
      --page-width <unitreal>         Page width
      --no-pdf-compression            Do not use lossless compression on pdf
                                      objects
  -q, --quiet                         Be less verbose
      --read-args-from-stdin          Read command line arguments from stdin
      --readme                        Output program readme
      --title <text>                  The title of the generated pdf file (The
                                      title of the first document is used if not
                                      specified)
      --use-xserver                   Use the X server (some plugins and other
                                      stuff might not work without X11)
  -V, --version                       Output version information and exit

Outline Options:
      --dump-default-toc-xsl          Dump the default TOC xsl style sheet to
                                      stdout
      --dump-outline <file>           Dump the outline to a file
      --outline                       Put an outline into the pdf (default)
      --no-outline                    Do not put an outline into the pdf
      --outline-depth <level>         Set the depth of the outline (default 4)

Page Options:
      --allow <path>                  Allow the file or files from the specified
                                      folder to be loaded (repeatable)
      --background                    Do print background (default)
      --no-background                 Do not print background
      --bypass-proxy-for <value>      Bypass proxy for host (repeatable)
      --cache-dir <path>              Web cache directory
      --checkbox-checked-svg <path>   Use this SVG file when rendering checked
                                      checkboxes
      --checkbox-svg <path>           Use this SVG file when rendering unchecked
                                      checkboxes
      --cookie <name> <value>         Set an additional cookie (repeatable),
                                      value should be url encoded.
      --custom-header <name> <value>  Set an additional HTTP header (repeatable)
      --custom-header-propagation     Add HTTP headers specified by
                                      --custom-header for each resource request.
      --no-custom-header-propagation  Do not add HTTP headers specified by
                                      --custom-header for each resource request.
      --debug-javascript              Show javascript debugging output
      --no-debug-javascript           Do not show javascript debugging output
                                      (default)
      --default-header                Add a default header, with the name of the
                                      page to the left, and the page number to
                                      the right, this is short for:
                                      --header-left='[webpage]'
                                      --header-right='[page]/[toPage]' --top 2cm
                                      --header-line
      --encoding <encoding>           Set the default text encoding, for input
      --disable-external-links        Do not make links to remote web pages
      --enable-external-links         Make links to remote web pages (default)
      --disable-forms                 Do not turn HTML form fields into pdf form
                                      fields (default)
      --enable-forms                  Turn HTML form fields into pdf form fields
      --images                        Do load or print images (default)
      --no-images                     Do not load or print images
      --disable-internal-links        Do not make local links
      --enable-internal-links         Make local links (default)
  -n, --disable-javascript            Do not allow web pages to run javascript
      --enable-javascript             Do allow web pages to run javascript
                                      (default)
      --javascript-delay <msec>       Wait some milliseconds for javascript
                                      finish (default 200)
      --keep-relative-links           Keep relative external links as relative
                                      external links
      --load-error-handling <handler> Specify how to handle pages that fail to
                                      load: abort, ignore or skip (default
                                      abort)
      --load-media-error-handling <handler> Specify how to handle media files
                                      that fail to load: abort, ignore or skip
                                      (default ignore)
      --disable-local-file-access     Do not allowed conversion of a local file
                                      to read in other local files, unless
                                      explicitly allowed with --allow
      --enable-local-file-access      Allowed conversion of a local file to read
                                      in other local files. (default)
      --minimum-font-size <int>       Minimum font size
      --exclude-from-outline          Do not include the page in the table of
                                      contents and outlines
      --include-in-outline            Include the page in the table of contents
                                      and outlines (default)
      --page-offset <offset>          Set the starting page number (default 0)
      --password <password>           HTTP Authentication password
      --disable-plugins               Disable installed plugins (default)
      --enable-plugins                Enable installed plugins (plugins will
                                      likely not work)
      --post <name> <value>           Add an additional post field (repeatable)
      --post-file <name> <path>       Post an additional file (repeatable)
      --print-media-type              Use print media-type instead of screen
      --no-print-media-type           Do not use print media-type instead of
                                      screen (default)
  -p, --proxy <proxy>                 Use a proxy
      --radiobutton-checked-svg <path> Use this SVG file when rendering checked
                                      radiobuttons
      --radiobutton-svg <path>        Use this SVG file when rendering unchecked
                                      radiobuttons
      --resolve-relative-links        Resolve relative external links into
                                      absolute links (default)
      --run-script <js>               Run this additional javascript after the
                                      page is done loading (repeatable)
      --disable-smart-shrinking       Disable the intelligent shrinking strategy
                                      used by WebKit that makes the pixel/dpi
                                      ratio none constant
      --enable-smart-shrinking        Enable the intelligent shrinking strategy
                                      used by WebKit that makes the pixel/dpi
                                      ratio none constant (default)
      --stop-slow-scripts             Stop slow running javascripts (default)
      --no-stop-slow-scripts          Do not Stop slow running javascripts
      --disable-toc-back-links        Do not link from section header to toc
                                      (default)
      --enable-toc-back-links         Link from section header to toc
      --user-style-sheet <url>        Specify a user style sheet, to load with
                                      every page
      --username <username>           HTTP Authentication username
      --viewport-size <>              Set viewport size if you have custom
                                      scrollbars or css attribute overflow to
                                      emulate window size
      --window-status <windowStatus>  Wait until window.status is equal to this
                                      string before rendering page
      --zoom <float>                  Use this zoom factor (default 1)

Headers And Footer Options:
      --footer-center <text>          Centered footer text
      --footer-font-name <name>       Set footer font name (default Arial)
      --footer-font-size <size>       Set footer font size (default 12)
      --footer-html <url>             Adds a html footer
      --footer-left <text>            Left aligned footer text
      --footer-line                   Display line above the footer
      --no-footer-line                Do not display line above the footer
                                      (default)
      --footer-right <text>           Right aligned footer text
      --footer-spacing <real>         Spacing between footer and content in mm
                                      (default 0)
      --header-center <text>          Centered header text
      --header-font-name <name>       Set header font name (default Arial)
      --header-font-size <size>       Set header font size (default 12)
      --header-html <url>             Adds a html header
      --header-left <text>            Left aligned header text
      --header-line                   Display line below the header
      --no-header-line                Do not display line below the header
                                      (default)
      --header-right <text>           Right aligned header text
      --header-spacing <real>         Spacing between header and content in mm
                                      (default 0)
      --replace <name> <value>        Replace [name] with value in header and
                                      footer (repeatable)

TOC Options:
      --disable-dotted-lines          Do not use dotted lines in the toc
      --toc-header-text <text>        The header text of the toc (default Table
                                      of Contents)
      --toc-level-indentation <width> For each level of headings in the toc
                                      indent by this length (default 1em)
      --disable-toc-links             Do not link from toc to sections
      --toc-text-size-shrink <real>   For each level of headings in the toc the
                                      font is scaled by this factor (default
                                      0.8)
      --xsl-style-sheet <file>        Use the supplied xsl style sheet f`

elicoidal commented on report_paper_wkhtmltopdf_params/README.rst:
```diff
@@ -0,0 +1,65 @@
+Contributors
+------------
+
+* Miku Laitinen <miku@avoin.systems>
```
contribution for the original author

JordiBForgeFlow replied:
He is the original author

elicoidal replied:
:+1:

elicoidal commented on report_paper_wkhtmltopdf_params/models/report.py:
```diff
@@ -0,0 +1,27 @@
+
+    def _build_wkhtmltopdf_args(self, paperformat,
+                                specific_paperformat_args=None):
+        # noinspection PyUnresolvedReferences,PyProtectedMember
```
not clear: remove or improve

JordiBForgeFlow replied:
This is actually very interesting, when you work with PyCharm, Will remove warning messages.

elicoidal commented on report_paper_wkhtmltopdf_params/models/report_paperformat_parameter.py:
```diff
@@ -0,0 +1,26 @@
+    name = fields.Char(
+        'Name',
+        required=True,
+        help='The command argument name. Remember to add prefix -- or -'
```
back to my previous comment, there is no control of the input: flexible but dangerous.

mlaitinen replied:
What kind of validation would you suggest?

JordiBForgeFlow replied:
I propose that we add not validation. The documentation provides enough details on how to properly set it up.

elicoidal reviewed (changes requested):
Please reconsider the scope of the module features

JordiBForgeFlow committed:
[fix] removed the field disable-smart-shrinking

JordiBForgeFlow commented:
@mlaitinen @elicoidal I have removed the field `disable_smart_shrinking`, and updated the documentation to use as an example how to add the parameter --disable-smart-shrinking. This makes the module generic for anyone to use it the way they want. 

I have tested and works for me very well.

pedrobaeza commented:
A possible validation is to call from @api.constrains method the command line with that arguments over a sample blank HTML file provided with this module. This way, you have dynamic argument checking depending on the wkhtmltopdf version you deploy in your system.

JordiBForgeFlow committed:
add constraint to verify the correctness of the parameters

JordiBForgeFlow commented:
@pedrobaeza thanks for the suggestion. I have implemented a small html. When the parameters are wrong it will not render a PDF, and then we raise an error to the user.

LoisRForgeFlow reviewed (approved):
Good one! :+1:

JordiBForgeFlow committed:
add tests

dreispt commented:
I wonder if this shouldn't rather be hosted at OCA/reporting-engine ?

JordiBForgeFlow commented:
@dreispt Yes, would be better there... @pedrobaeza your thoughts?

pedrobaeza commented:
I agree, and renamed without "paper" part and without plural: `report_wkhtmltopdf_param`

JordiBForgeFlow commented:
@pedrobaeza You kill me, man.... I have this already deployed in a production environment.

pedrobaeza commented:
hehe, but it's easy from a shell command:
```
from openupgradelib import openupgrade

openupgrade.update_module_names([('report_paper_wkhtmltopdf_params', 'report_wkhtmltopdf_param')])
```

Another option is to rename column, uninstall module, rename again, and install the new one.

You already knew before the no plural in module names convention...

JordiBForgeFlow commented:
ok... will do that

dreispt commented:
Can't you bump module version and provide a migration script in the module? (Not that I have done that myself...)

pedrobaeza commented:
Not for module renamings.
